sdk-extension/opentelemetry-sdk-extension-aws/src/opentelemetry/sdk/extension/aws/trace/propagation/aws_xray_format.py

```diff
@@ -0,0 +1,284 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import typing
+
+import opentelemetry.trace as trace
+from opentelemetry.context import Context
+from opentelemetry.trace.propagation.textmap import (
+    Getter,
+    Setter,
+    TextMapPropagator,
+    TextMapPropagatorT,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class AwsXRayFormat(TextMapPropagator):
+    """Propagator for the AWS X-Ray Trace Header propagation protocol.
+
+    See: https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader
+    """
+
+    # AWS
+    TRACE_HEADER_KEY = "X-Amzn-Trace-Id"
+
+    KV_PAIR_DELIMITER = ";"
+    KEY_AND_VALUE_DELIMITER = "="
+
+    TRACE_ID_KEY = "Root"
+    TRACE_ID_LENGTH = 35
+    TRACE_ID_VERSION = "1"
+    TRACE_ID_DELIMITER = "-"
+    TRACE_ID_DELIMITER_INDEX_1 = 1
+    TRACE_ID_DELIMITER_INDEX_2 = 10
+    TRACE_ID_FIRST_PART_LENGTH = 8
+
+    PARENT_ID_KEY = "Parent"
+    PARENT_ID_LENGTH = 16
+
+    SAMPLED_FLAG_KEY = "Sampled"
+    SAMPLED_FLAG_LENGTH = 1
+    IS_SAMPLED = "1"
+    NOT_SAMPLED = "0"
+
+    # pylint: disable=too-many-locals
+    # pylint: disable=too-many-return-statements
+    # pylint: disable=too-many-branches
+    # pylint: disable=too-many-statements
+    def extract(
+        self,
+        getter: Getter[TextMapPropagatorT],
+        carrier: TextMapPropagatorT,
+        context: typing.Optional[Context] = None,
+    ) -> Context:
+        trace_header_list = getter(carrier, self.TRACE_HEADER_KEY)
+
+        if not trace_header_list or len(trace_header_list) != 1:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_header = trace_header_list[0]
+
+        if not trace_header:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_id = trace.INVALID_TRACE_ID
+        span_id = trace.INVALID_SPAN_ID
+        sampled = False
+
+        next_kv_pair_start = 0
+
+        while next_kv_pair_start < len(trace_header):
+            try:
+                kv_pair_delimiter_index = trace_header.index(
```

Comment:
this parsing logic can be done a bit simpler, and results in operations that are the same (or cheaper) than what's written here:

```
for component in tracer_header.split(self.KV_PAIR_DELIMITER):
    try:
        key, value = component.split(self.KEY_AND_VALUE_DELIMITER)
        if key == self.TRACE_ID_KEY:
            # trace_id_key logic
        else ...
    except ValueError:
       # return invalid tracecontext / return error value
```

In python, every index slice created results in a brand new string, so the current approach is pretty expensive because it's creating a bunch of sub-strings, then further slicing them. The split() function in Python IIRC is written in C, so it'll be much faster than any code you could write in Python itself.

Reply:
I really like this. Most of the motivation I had for writing it the way I did was to make sure I was following [what java did](https://github.com/open-telemetry/opentelemetry-java/blob/3014bb7fa05f6314e594a187cd98d48524ae17f6/extensions/trace-propagators/src/main/java/io/opentelemetry/extension/trace/propagation/AwsXRayPropagator.java) because I knew it worked. Specifically, I was wondering if there were some HTTP header specifics that I didn't know about... but after adding all the tests I feel comfortable this works even in a variety of situations. So I'll make the changes because I think it does read (and should run like you mentioned) better!